src/apps/app_bulk_upload.erl

```diff
@@ -0,0 +1,212 @@
+	}.
+
+chunk_tags(false) ->
+	[
+		{<< "app_name" >>, << "BulkUpload" >>}
+	];
+chunk_tags(true) ->
+	chunk_tags(false) ++ [{<< "first_chunk" >>, << "true" >>}].
+
+%% @doc Searches the local storage for the chunks of the blob identified by the given
+%% Base 64 encoded transaction ID. If the blob is reconstructed successfully,
+%% writes it to the specified destination.
+download(TXID, Filename) ->
+	{ok, Blob} = download(TXID),
+	file:write_file(Filename, Blob, [write]).
+
+download(TXID) ->
+	case download_chunks(TXID, []) of
+		Chunks when is_list(Chunks) ->
+			{ok, Chunks};
+		Err ->
+			{error, Err}
+	end.
+
+download_chunks(TXID, Chunks) ->
+	case ar_storage:read_tx(ar_util:decode(TXID)) of
+		unavailable ->
+			tx_not_found;
+		TX ->
+			case lists:keyfind(<< "app_name" >>, 1, TX#tx.tags) of
+				false ->
+					invalid_tx;
+				_ ->
+					case lists:keyfind(<< "first_chunk" >>, 1, TX#tx.tags) of
+						false ->
+							download_chunks(ar_util:encode(TX#tx.last_tx), [TX#tx.data|Chunks]);
+						_ ->
+							[TX#tx.data|Chunks]
+					end
+			end
+	end.
+
+upload_one_chunk_test_() ->
+	{
+		timeout, 60, fun() ->
+			upload_blob_test_with_blob_size(?CHUNK_SIZE - 1)
+		end
+	}.
+
+upload_one_even_chunk_test_() ->
+	{
+		timeout, 60, fun() ->
+			upload_blob_test_with_blob_size(?CHUNK_SIZE)
+		end
+	}.
+
+upload_two_even_chunks_test_() ->
+	{
+		timeout, 60, fun() ->
+			upload_blob_test_with_blob_size(?CHUNK_SIZE * 2)
+		end
+	}.
+
+upload_three_chunks_test_() ->
+	{
+		timeout, 60, fun() ->
+			upload_blob_test_with_blob_size(?CHUNK_SIZE * 2 + 1)
+		end
+	}.
+
+upload_blob_test_with_blob_size(BlobSize) ->
+	Wallet = {_, Pub} = ar_wallet:new(),
+	Bs = ar_weave:init([{ar_wallet:to_address(Pub), ?AR(10000), <<>>}]),
+	Node = ar_node:start([], Bs),
+
+	Blob = iolist_to_binary(generate_blob(BlobSize)),
+
+	upload_blob(Node, Wallet, Blob),
+
+	BlobSizeByChunkSize = BlobSize div ?CHUNK_SIZE,
+	Remainder = case BlobSize rem ?CHUNK_SIZE of 0 -> 0; _ -> 1 end,
+	ExpectedTXNumber = BlobSizeByChunkSize + Remainder,
+
+	mine_blocks(Node, ExpectedTXNumber * 3),
+	Transactions = collect_transactions(Node, Wallet),
+	?assertEqual(ExpectedTXNumber, length(Transactions)),
+	?assertEqual(hash(Blob), hash(<< (TX#tx.data) || TX <- Transactions >>)),
+	[First|Rest] = Transactions,
+	?assertEqual(
+		[
+			{<< "app_name" >>, << "BulkUpload" >>},
+			{<< "first_chunk" >>, << "true" >>}
+		],
+		First#tx.tags
+	),
+	assert_rest(Rest),
+	LastTX = lists:last(Transactions),
+	{ok, DownloadedChunks} = download(ar_util:encode(LastTX#tx.id)),
+	?assertEqual(hash(Blob), hash(iolist_to_binary(DownloadedChunks))).
+
+hash(Blob) ->
+	ar_util:encode(crypto:hash(?BLOB_HASH_ALGO, Blob)).
+
+%% Generates an iolist() of the given size.
+%% The first 100 bytes are randomly picked from [1; 100].
+%% The rest is filled with 1. It would take too long to randomly pick every byte for a huge blob.
+generate_blob(Size) ->
+	generate_blob(Size, []).
+
+generate_blob(0, Acc) ->
```

Comment:
I don't understand this function. How about just generate a binary with `Size` number of random bytes?

E.g (non-tested):

```
random_blob(Size) ->
    random_blob(Size, []).

random_blob(Size, Acc) when length(Acc) == Size ->
    list_to_binary(Acc);
random_blob(Size, Acc) ->
    Byte = rand:uniform(255),
    random_blob([Byte | Acc]).
```

Reply:
Oh yes, that's 2 MB of random data which takes a while. But actually, the blobs doesn't have to be random in the test since the size of the three blobs are different. It would also be good to test an uploading and downloading the same binary twice.

Reply:
The blobs have to be random cause there is no cleanup in between test runs.